src/lsstseries/ensemble.py

```diff
@@ -391,6 +408,69 @@ def batch(self, func, *args, meta=None, use_map=True, compute=True, on=None, **k
         else:
             return batch
 
+    def from_hipscat(
+        self,
+        dir,
+        source_subdir="source",
+        object_subdir="object",
+        id_col=None,
+        time_col=None,
+        flux_col=None,
+        err_col=None,
+        band_col=None,
+        additional_cols=True,
+    ):
+        """Read in parquet files from a hipscat-formatted directory structure
+        Parameters
+        ----------
+        dir: 'str'
+            Path to the directory structure
+        source_subdir: 'str'
+            Path to the subdirectory which contains source files
+        object_subdir: 'str'
+            Path to the subdirectory which contains object files, if None then
+            files will only be read from the source_subdir
+        id_col: 'str', optional
+            Identifies which column contains the Object IDs
+        time_col: 'str', optional
+            Identifies which column contains the time information
+        flux_col: 'str', optional
+            Identifies which column contains the flux/magnitude information
+        err_col: 'str', optional
+            Identifies which column contains the flux/mag error information
+        band_col: 'str', optional
+            Identifies which column contains the band information
+        additional_cols: 'bool', optional
+            Boolean to indicate whether to carry in columns beyond the
+            critical columns, true will, while false will only load the columns
+            containing the critical quantities (id,time,flux,err,band)
+
+        Returns
+        ----------
+        ensemble: `lsstseries.ensemble.Ensemble`
+            The ensemble object with parquet data loaded
+        """
+
+        source_path = os.path.join(dir, source_subdir)
```

Comment:
Thanks for this comment, it has really got me thinking about what’s missing from this current implementation. Sorry for the incoming brain dump.

For one, you’re definitely correct in that we aren’t utilizing the association information provided by hipscat. We don’t in principle care about this for the actual loading of the data in, as we’re just building two separate dataframes with trust that there’s a common id column to join them in later operations, the only immediate join/merge happening is between the object table and an object table built off the source table. Which, short aside, we don’t have to do this, as the main purpose of it right now is to populate count columns (nobs_total, nobs_r, etc.). We can alternatively just trust those columns as provided by the object table. But where we’ll suffer currently is once we have the data in our dataframes, any time we need to do joins, dask will have to figure out the association as we have no information available for it. 

It’s still not that clear to me how we would use the available hipscat information to supply this to dask dataframes, but one solution to this is to utilize dask dataframe divisions. Which, per partition, denote the minimum and maximum values of the index found in each partition (requires a sorted index). This gets at what you’re suggesting, where it would enable joins to know which partitions in source a given index in object would live. We probably shouldn’t use _id_col as the index in this case, as we’d have to sort it (clobbering any smart partitioning scheme). But we could enumerate the id_col, and use that sorted integer index as our lsstseries index. We can generate that index all within lsstseries, but we still wouldn’t be actually utilizing any information from hipscat. I would have liked to use something like the hipscat_index for this, but the counter is unique per-source, where we need something unique per-object, and mapped to all sources (non-unique). Is there something where we can use the association table to generate an index that is then passed along to the object and source tables? 

The fact that the data format is not set by lsdb/hipscat is actually news to me. Support for alternative data formats is something we should do…in the future.

Reply:
I tried to refactor lsstseries to allow the type of association information that hipscat provides (logged in #98), but ended up with something ~2-5x slower, at least at the scale of 1 million lightcurves or less. At present, we don't do a ton of joins between object and source, really only when doing sync operations between the dataframes. It sounds like potentially what we could do, is use the localized join functionality of LSDB once that has made it into the codebase to generate our object and source tables. This would give us a source table with sources co-located by object id, which for our purposes is ideal.